add implementation considerations section to address gap between serialization format and APIs in FIPS 203 (#79)
lamps-wg/kyber-certificates: draft-ietf-lamps-kyber-certificates.md

{{example-private}} contains examples for ML-KEM private keys
encoded using the textual encoding defined in {{?RFC7468}}.

# Implementation Considerations

Though section 7.1 of {{FIPS203}} mentions the potential to save seed values for future expansion, Algorithm 19 does not make the seed values available to a caller for serialization.

> Similarly, the algorithm that expands seed values is not listed as one of the "main algorithms" and features "internal" in the name.

Reviewer: ```suggestion
Similarly, the algorithm that expands seed values is not listed as one of the "main algorithms" and features "internal" in the name even though it is clear that it is allowed to be exposed externally for the purposes of expanding a key from a seed.
```

Below are possible ways to extend the APIs defined in {{FIPS203}} to support serialization of seed values as private keys.

To support serialization of seed values as private keys, Algorithm 19 in {{FIPS203}} should return (ek, dk, d, z) on line 7 and Algorithm 16 should be promoted to be a "main algorithm" for external use in expanding seed values.

> 

Reviewer: ```suggestion

Note also that unlike other private key compression methods in other algorithms, expanding a private key from a seed is a one-way function, meaning that once a full key is expanded from seed and the seed discarded, the seed cannot be re-created even if the full expanded private key is available. For this reason it is RECOMMENDED that implementations retain and export the seed, even when also exporting the expanded key.
```

{{example-private}} contains examples for ML-KEM private keys
encoded using the textual encoding defined in {{?RFC7468}}.

# Implementation Considerations

Though section 7.1 of {{FIPS203}} mentions the potential to save seed values for future expansion, Algorithm 19 does not make the seed values available to a caller for serialization.
Similarly, the algorithm that expands seed values is not listed as one of the "main algorithms" and features "internal" in the name even though it is clear that it is allowed to be exposed externally for the purposes of expanding a key from a seed.
Below are possible ways to extend the APIs defined in {{FIPS203}} to support serialization of seed values as private keys.

To support serialization of seed values as private keys, Algorithm 19 in {{FIPS203}} should return (ek, dk, d, z) on line 7 and Algorithm 16 should be promoted to be a "main algorithm" for external use in expanding seed values.

> Note also that unlike other private key compression methods in other algorithms, expanding a private key from a seed is a one-way function, meaning that once a full key is expanded from seed and the seed discarded, the seed cannot be re-created even if the full expanded private key is available. For this reason it is RECOMMENDED that implementations retain and export the seed, even when also exporting the expanded key.

Reviewer 2: ```suggestion
Note also that unlike other private key compression methods in other algorithms, expanding a private key from a seed is a one-way function, meaning that once a full key is expanded from seed and the seed discarded, the seed cannot be re-created even if the full expanded private key is available. For this reason it is RECOMMENDED that implementations retain and export the seed, even when also exporting the expanded key.

```

{{example-private}} contains examples for ML-KEM private keys
encoded using the textual encoding defined in {{?RFC7468}}.

# Implementation Considerations

Though section 7.1 of {{FIPS203}} mentions the potential to save seed values for future expansion, Algorithm 19 does not make the seed values available to a caller for serialization.
Similarly, the algorithm that expands seed values is not listed as one of the "main algorithms" and features "internal" in the name even though it is clear that it is allowed to be exposed externally for the purposes of expanding a key from a seed.
Below are possible ways to extend the APIs defined in {{FIPS203}} to support serialization of seed values as private keys.

> To support serialization of seed values as private keys, Algorithm 19 in {{FIPS203}} should return (ek, dk, d, z) on line 7 and Algorithm 16 should be promoted to be a "main algorithm" for external use in expanding seed values.

Reviewer 3: > Algorithm 19 in {{FIPS203}} should return

This sounds like we're redefining FIPS203 instead of extending it. Better to phrase it like let "alg 19b" denote the same procedure as algorithm 19 except that it returns (ek, dk, d, z) on line 7.

Author: Works for me.

Reviewer 2: @carl-wallace are you going to suggest some changes I can merge in?

Author: Applying Bas' suggestion to the PR I sent yields something like this:

To support serialization of seed values as private keys, let Algorithm 19b denote the same procedure as Algorithm 19 in {{FIPS203}} except it returns (ek, dk, d, z) on line 7. Additionally, Algorithm 16 should be promoted to be a "main algorithm" for external use in expanding seed values.

> To support serialization of seed values as private keys, Algorithm 19 in {{FIPS203}} should return (ek, dk, d, z) on line 7 and Algorithm 16 should be promoted to be a "main algorithm" for external use in expanding seed values.

Reviewer 2: ```suggestion
To support serialization of seed values as private keys, let Algorithm 19b denote the same procedure as Algorithm 19 in {{FIPS203}} except it returns (ek, dk, d, z) on line 7. Additionally, Algorithm 16 should be promoted to be a "main algorithm" for external use in expanding seed values.
```
